cmd/kubeadm/app/cmd/join.go

```diff
@@ -303,6 +303,59 @@ func newJoinOptions() *joinOptions {
 // This func takes care of validating joinOptions passed to the command, and then it converts
 // options into the internal JoinConfiguration type that is used as input all the phases in the kubeadm join workflow
 func newJoinData(cmd *cobra.Command, args []string, opt *joinOptions, out io.Writer) (*joinData, error) {
+
+	// amend the values in joinOptions
+	amendJoinOptions(opt, args)
+
+	// if the admin.conf file already exists, use it for skipping the discovery process.
+	// NB. this case can happen when we are joining a control-plane node only (and phases are invoked atomically)
+	var adminKubeConfigPath = kubeadmconstants.GetAdminKubeConfigPath()
+	var tlsBootstrapCfg *clientcmdapi.Config
+	if _, err := os.Stat(adminKubeConfigPath); err == nil && opt.controlPlane {
+		// use the admin.conf as tlsBootstrapCfg, that is the kubeconfig file used for reading the kubeadm-config during discovery
+		klog.V(1).Infof("[preflight] found %s. Use it for skipping discovery", adminKubeConfigPath)
+		tlsBootstrapCfg, err = clientcmd.LoadFromFile(adminKubeConfigPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error loading %s", adminKubeConfigPath)
+		}
+	}
+
+	ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(opt.ignorePreflightErrors)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = validation.ValidateMixedArguments(cmd.Flags()); err != nil {
+		return nil, err
+	}
+
+	// in case the command doesn't have flags for discovery, makes the join cfg validation pass checks on discovery
+	if cmd.Flags().Lookup(options.FileDiscovery) == nil {
+		if _, err := os.Stat(adminKubeConfigPath); os.IsNotExist(err) {
+			return nil, errors.Errorf("File %s does not exists. Please use 'kubeadm join phase control-plane-prepare' subcommands to generate it.", adminKubeConfigPath)
+		}
+		klog.V(1).Infof("[preflight] found discovery flags missing for this command. using FileDiscovery: %s", adminKubeConfigPath)
+		opt.externalcfg.Discovery.File = &kubeadmapiv1beta1.FileDiscovery{KubeConfigPath: adminKubeConfigPath}
+		opt.externalcfg.Discovery.BootstrapToken = nil //NB. this could be removed when we get better control on args (e.g. phases without discovery should have NoArgs )
+	}
+
+	cfg, err := loadJoinConfiguration(opt)
+	if err != nil {
+		return nil, err
+	}
+
+	return &joinData{
+		cfg:                   cfg,
+		tlsBootstrapCfg:       tlsBootstrapCfg,
+		ignorePreflightErrors: ignorePreflightErrorsSet,
+		outputWriter:          out,
+		certificateKey:        opt.certificateKey,
+	}, nil
+}
+
+// amendJoinOptions defaults the kubeadm configuration and applies modifications to it
+// based on the passed arguments.
+func amendJoinOptions(opt *joinOptions, args []string) {
```

Comment:
I don't feel that this is the best name for the func. It should be something on the lines of `amendExternalJoinConfiguration` in order to clarify that it's mainly `opt.externalcfg` that is being patched with command line args.